packages/cli/src/controller/scaffoldgen-controller.ts

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import fs from 'fs';
+import path from 'path';
+import {EventFragment, FunctionFragment} from '@ethersproject/abi/src.ts/fragments';
+import {loadFromJsonOrYaml} from '@subql/common';
+import ejs from 'ejs';
+import {Interface} from 'ethers/lib/utils';
+import * as inquirer from 'inquirer';
+import {upperFirst} from 'lodash';
+import {parseContractPath} from 'typechain';
+import {parseDocument} from 'yaml';
+import {SelectedMethod, UserInput} from '../commands/codegen/generate';
+
+const SCAFFOLD_HANDLER_TEMPLATE_PATH = path.resolve(__dirname, '../template/scaffold-handlers.ts.ejs');
+const ROOT_MAPPING_DIR = 'src/mappings';
+const DEFAULT_HANDLER_BUILD_PATH = './dist/index.js';
+
+interface TopicsFilter {
+  topics: string[];
+}
+interface FunctionFilter {
+  function: string;
+}
+type Filter = TopicsFilter | FunctionFilter;
+
+interface HandlerType {
+  handler: string;
+  kind: string;
+  filter: Filter;
+}
+
+interface DatasourceProp {
+  kind: string;
+  startBlock: number;
+  options: {
+    abi: string;
+    address?: string;
+  };
+  assets: {
+    [key: string]: {
+      file: string;
+    };
+  };
+  mapping: {
+    file: string;
+    handlers: HandlerType[];
+  };
+}
+
+export async function promptSelectables(
+  input: string,
+  availableMethods: string[],
+  memArr: string[]
+): Promise<string[]> {
+  if (input === '*') {
+    return availableMethods;
+  }
+
+  if (input) {
+    try {
+      const chosenFn = await inquirer.prompt({
+        name: 'functions',
+        message: 'Select Functions',
+        type: 'checkbox',
+        choices: availableMethods,
+      });
+      memArr.push(...chosenFn.functions);
+    } catch (e) {
+      throw new Error(e);
+    }
+  }
+  return memArr;
+}
+
+export async function renderTemplate(templatePath: string, outputPath: string, templateData: ejs.Data): Promise<void> {
+  const data = await ejs.renderFile(templatePath, templateData);
+  await fs.promises.writeFile(outputPath, data);
+}
+
+export function getAbiInterface(projectPath: string, abiPath: string): Interface {
+  const abi = loadFromJsonOrYaml(path.join(projectPath, abiPath)) as any;
+  return new Interface(abi);
+}
+
+export function getAvailableEvents(abiInterface: Interface): {[p: string]: EventFragment} {
```

Comment:
`Record` is more readable 
```suggestion
export function getAvailableEvents(abiInterface: Interface): Record<string, EventFragment> {
```

Reply:
Also do you think this function is necessary?